components/automate-cli/pkg/verifyserver/services/certificatevalidation/certificatevalidation.go

```diff
@@ -0,0 +1,252 @@
+		Title:         title,
+		Passed:        passed,
+		SuccessMsg:    successMsg,
+		ErrorMsg:      errorMsg,
+		ResolutionMsg: resolutionMsg,
+	}
+}
+
+func createErrorMessage(expiredCerts, aboutToExpireCerts string) string {
+	errorMessage := ""
+	if expiredCerts != "" {
+		expiredCerts = strings.TrimSuffix(expiredCerts, ", ")
+		errorMessage = fmt.Sprintf(constants.CERTIFICATE_EXPIRY_ERROR_MESSAGE, expiredCerts)
+	}
+	if aboutToExpireCerts != "" {
+		aboutToExpireCerts = strings.TrimSuffix(aboutToExpireCerts, ", ")
+		if errorMessage != "" {
+			errorMessage += "; "
+		}
+		errorMessage += fmt.Sprintf(constants.CERTIFICATE_INVALID_EXPIRY_MESSAGE, aboutToExpireCerts)
+	}
+
+	return errorMessage
+}
+
+func decodeAndParseCertificate(certificate, key string) (*x509.Certificate, error) {
+	block, _ := pem.Decode([]byte(certificate))
+	if block == nil || block.Type != "CERTIFICATE" {
+		return nil, fmt.Errorf("failed to decode %s certificate", key)
+	}
+
+	parsedCertificate, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse %s certificate: %s", key, err.Error())
+	}
+
+	return parsedCertificate, nil
+}
+
+func (vc *ValidateCertificateService) validateCertificateExpiry(certificates map[string]string) models.Checks {
+	vc.log.Debug("Validating Certificates Expiry...")
+	expiredCerts := ""
+	aboutToExpireCerts := ""
+	keys := []string{"Root", "Node", "Admin"}
+
+	for _, key := range keys {
+		cert := certificates[key]
+		certificate, err := decodeAndParseCertificate(cert, key)
+		if err != nil {
+			vc.log.Error(err)
+			expiredCerts += key
+			expiredCerts += ", "
+			continue
+		}
+
+		// this is for checking that certificate is expired or not.
+		currentTime := time.Now()
+		if currentTime.Before(certificate.NotBefore) || currentTime.After(certificate.NotAfter) {
+			vc.log.Debugf("%s certificate is expired", key)
+			expiredCerts += key
+			expiredCerts += ", "
+			continue
+		}
+
+		// this is for checking that certificate is going to expire in next 365 days or not.
+		expirationDate := time.Now().AddDate(0, 0, 365)
+		if certificate.NotAfter.Before(expirationDate) {
+			vc.log.Debugf("%s certificate is going to expire in next 365 days", key)
+			aboutToExpireCerts += key
+			aboutToExpireCerts += ", "
+			continue
+		}
+	}
+
+	if expiredCerts == "" && aboutToExpireCerts == "" {
+		vc.log.Debug("Certificates are not expired")
+		return createCheck(constants.CERTIFICATE_EXPIRY_TITLE, true, constants.CERTIFICATE_EXPIRY_SUCCESS_MESSAGE, "", "")
+	}
+
+	errorMessage := createErrorMessage(expiredCerts, aboutToExpireCerts)
+	return createCheck(constants.CERTIFICATE_EXPIRY_TITLE, false, "", errorMessage, constants.CERTIFICATE_EXPIRY_RESOLUTION_MESSAGE)
+}
+
+func (vc *ValidateCertificateService) validateCertificateFormat(certificates map[string]string) models.Checks {
+	vc.log.Debug("Validating Certificates Format...")
+	invalidFormatCerts := ""
+	keys := []string{"Root", "Node", "Admin"}
```

Comment:
Since, maps does not preserve the order of keys and i was getting different output every time, So to fix that and majorly for testing responses, i created separate array and traverse that.